scrape/rules.go

```diff
@@ -0,0 +1,197 @@
+// Copyright 2022 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scrape
+
+import (
+	"context"
+	"time"
+
+	"github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/storage"
+)
+
+type RuleEngine interface {
+	NewScrapeBatch() Batch
+	EvaluateRules(b Batch, ts time.Time) ([]*Sample, error)
+}
+
+// ruleEngine evaluates rules from individual targets at scrape time.
+type ruleEngine struct {
+	targetLabels labels.Labels
+	rules        []*config.ScrapeRuleConfig
+	engine       *promql.Engine
+}
+
+// newRuleEngine creates a new RuleEngine.
+func newRuleEngine(targetLabels labels.Labels, rules []*config.ScrapeRuleConfig) RuleEngine {
+	if len(rules) == 0 {
+		return &nopRuleEngine{}
+	}
+
+	return &ruleEngine{
+		targetLabels: targetLabels,
+		rules:        rules,
+		engine: promql.NewEngine(promql.EngineOpts{
+			MaxSamples:    50000000,
+			Timeout:       10 * time.Second,
+			LookbackDelta: 5 * time.Minute,
+		}),
+	}
+}
+
+// NewScrapeBatch creates a new Batch which can be used to add samples from a single scrape.
+// Rules are always evaluated on a single Batch.
+func (r *ruleEngine) NewScrapeBatch() Batch {
+	return &batch{
+		samples: make([]*Sample, 0),
+	}
+}
+
+// EvaluateRules executes rules on the given Batch and returns new Samples.
+func (r *ruleEngine) EvaluateRules(b Batch, ts time.Time) ([]*Sample, error) {
+	var result []*Sample
+	for _, rule := range r.rules {
+		queryable := storage.QueryableFunc(func(ctx context.Context, mint, maxt int64) (storage.Querier, error) {
+			return b, nil
+		})
+
+		query, err := r.engine.NewInstantQuery(queryable, nil, rule.Expr, ts)
+		if err != nil {
+			return nil, err
+		}
+
+		samples, err := query.Exec(context.Background()).Vector()
+		if err != nil {
+			return nil, err
+		}
+
+		for _, s := range samples {
+			lbls := s.Metric.WithoutLabels("__name__")
```

Comment:
You can use `labels.MetricName` instead of hardcoding `__name__` here.

https://github.com/prometheus/prometheus/blob/main/model/labels/labels.go#L27

Reply:
Good point, updated.